### ISSUE TITLE
In order to use an SSH DOCKER_HOST on OS X and Linux, you must configure an ssh-agent.

### ISSUE THREAD

**Issue by maxsatula** [microsoft/vscode-docker#1459]
https://github.com/microsoft/vscode-docker/blob/836a76500c1e3edaadccdea944dec31b3e586c2c/src/utils/refreshDockerode.ts#L63

https://github.com/microsoft/vscode-docker/blob/836a76500c1e3edaadccdea944dec31b3e586c2c/src/utils/refreshDockerode.ts#L81

I doubt setting of `SSH_AUTH_SOCK` variable is necessary. I keep it undefined, but have `IdentityAgent`, `IdentityFile`, `AddKeysToAgent` in my `.ssh/config`. However the code referenced at the top makes my configuration unusable for this extension, while it perfectly works everywhere else outside vscode-docker.

Can this check be removed or replaced with something else? There are other options for SSH to avoid interactive password prompt, such as `BatchMode` or `IdentitiesOnly`.

Thank you!

**Comment by philliphoff** [comment]
Some of the manipulation of environment variables related to SSH is due limitations in the way `dockerode` (used to connect to the Docker host) must be initialized.  That said, there may be more that we could do to support a wider variety of SSH configurations.

**Comment by maxsatula** [comment]
Yep, and dockerode does not seem to respect `.ssh/config` file at all, so the solution might not be that simple as I suggested earlier :(

**Comment by maxsatula** [comment]
Nonetheless, I would appreciate a possibility to customize value of `SSH_AUTH_SOCK`.
I run agent as `ssh-agent -a <my_fixed_path_to_UNIX_socket>`, so the path is known and static.

**Comment by bwateratmsft** [comment]
Dockerode indeed does not respect it; Dockerode is using the [ssh2](https://www.npmjs.com/package/ssh2) node package which does not use `.ssh/config`. I would suggest filing this there because it would be a better solution for it to be implemented there.

**Comment by maxsatula** [comment]
> Nonetheless, I would appreciate a possibility to customize value of `SSH_AUTH_SOCK`.

Do you think it makes sense to implement this part here?

**Comment by AVSuni** [comment]
I get the same error message on three separate laptops (one running Ubuntu 18.04 and two running Windows 10). I don't get it. I have followed all the instructions, and this used to work perfectly well until about October/November.

**Comment by lcxywfe** [comment]
> Dockerode indeed does not respect it, unfortunately; Dockerode is using the [ssh2](https://www.npmjs.com/package/ssh2) node package which does not use `.ssh/config`. I would suggest filing this there because it would be a better solution for it to be implemented there (in ssh2).

Sorry, so, does it support ssh with a jump host ?

**Comment by bwateratmsft** [comment]
@lcxywfe it may be possible to use a jump host but we have never tried it.

**Comment by lcxywfe** [comment]
@bwateratmsft I mean, could tell me how to modify the DOCKER_HOST (or another variable) to use a jump host.

**Comment by steverice** [comment]
It seems like ssh2 [already supports specifying an agent](https://github.com/mscdex/ssh2#client-methods), there's just no way to specify ssh2 options in vscode-docker and have them provided to the ssh2 client.

`docker-modem` will [pass the options it gets along to ssh2](https://github.com/apocas/docker-modem/blob/a4c7f806c124f7c1e59d2493d5a9382924724558/lib/ssh.js#L24), and `dockerode` will [pass along its options as well](https://github.com/apocas/dockerode/blob/a6f27b32cc2aed825f0afe18a16c7e6caf731e02/lib/docker.js#L32).

It seems like the missing link is just that [`vscode-docker` doesn't provide any config](https://github.com/apocas/dockerode/blob/a6f27b32cc2aed825f0afe18a16c7e6caf731e02/lib/docker.js#L32) when initializing `Dockerode`. If `vscode-docker` gave us a way to specify options for `ssh2` that would be used here, we should be able to use `ssh2` options like `agent` and `agentForward`.

**Comment by bwateratmsft** [comment]
@steverice this is already possible, it's what the `SSH_AUTH_SOCK` environment variable is designed for. No need for config in-product to do exactly the same thing, especially since it's a temporary socket on Mac and Linux that changes every reboot.

**Comment by maxsatula** [comment]
@bwateratmsft 
Could you please point out how SSH_AUTH_SOCK variable should be initialized when vscode runs on Windows, and it's Linux counterpart starts automatically through Remote WSL plugin? I basically was unable to find a right spot to set it.

Secondly, it is not always true that socket changes at every reboot. If agent runs as `ssh-agent -a <your_permanent_socket_path>`, then one can rely on a permanent location. So it does make sense (at least for those people who run ssh agent like that) to have a persistent configuration of what SSH_AUTH_SOCK is.

**Comment by bwateratmsft** [comment]
On Windows, you need to set up OpenSSH (more info [here](https://code.visualstudio.com/docs/containers/ssh#_set-up-ssh-tunneling)). On Windows with OpenSSH as the SSH agent, it's always `\\.\pipe\openssh-ssh-agent`. You can use Pageant instead of OpenSSH, in which case you must set `SSH_AUTH_SOCK` to `pageant`.

Because the path is always predictable on Windows, if `SSH_AUTH_SOCK` is unset, we assume it is `\\.\pipe\openssh-ssh-agent`. That code is [here](https://github.com/microsoft/vscode-docker/blob/master/src/utils/refreshDockerode.ts#L75).

For those that use the `ssh-agent -a` option--it's not any more difficult to use. You still do `eval $(ssh-agent ...)`, which starts the agent and sets the environment variables. Also, on at least Ubuntu, it wasn't necessary to do anything with `ssh-agent`, because it's done automatically, including setting `SSH_AUTH_SOCK`. All that was needed was a one-time `ssh-add <keyfile>`.

In any case, I'd like to point out that using an `ssh://` `DOCKER_HOST` or `docker.host`, while possible, is not the preferred way of using the extension with SSH. Using the Remote - SSH extension with the Docker extension in "workspace" mode and installed in the remote session is going to offer a better experience with less setup overhead. More info [here](https://code.visualstudio.com/docs/containers/choosing-dev-environment#_remote-machine).

**Comment by maxsatula** [comment]
You described the cases when vscode is used
1. On Windows, no WSL
2. On Linux natively

I was talking about WSL - Remote configuration, more info [here](https://marketplace.visualstudio.com/items?itemName=ms-vscode-remote.remote-wsl),
where Windows part is just used for rendering, and WSL part does all the work. vscode runs some sort of server on WSL in background and talks to it, and I am not sure I can set any environment variables upon vscode server startup (not the Windows part which would be easy as you described, but it does just rendering). Neither it is a standalone Linux execution, which otherwise would also be easy to setup as you described.

**Comment by bwateratmsft** [comment]
Just to make sure I understand your scenario correctly, are you using Remote - WSL to connect to WSL, but the Docker daemon is running somewhere else and you need to use SSH to get to it? Or is Docker also running in WSL?

**Comment by maxsatula** [comment]
Correct, docker is running elsewhere.

**Comment by bwateratmsft** [comment]
Gotcha. For WSL, at least the Ubuntu flavor, you can accomplish this by putting something like this into `/home/username/.bashrc`, probably toward the bottom:

```sh
if [ ! -e ~/mysock ]; then
  eval $(ssh-agent -s -a ~/mysock)
  ssh-add ~/id_rsa
else
  echo SSH agent already running
  export SSH_AUTH_SOCK=~/mysock
fi
```

What this does is, if the socket does not exist (first run of the session), it starts `ssh-agent` with the specified `~/mysock` as the `SSH_AUTH_SOCK` and adds the keyfile; otherwise if it does exist it simply sets the variable.

That `echo` in the `else` block is optional but made it easier to tell what it was doing.

(P.S. I'm eagerly awaiting WSL2 myself. :smile: It's a huge improvement over the local VM that Windows does today, particularly in performance.)

**Comment by maxsatula** [comment]
You gonna kill me now...
Sure, it occurred to me to set something in rc file, rather than opening an issue here, but it affects entire WSL environment, which is used for other stuff, thus is not considered a clean solution. Say, I have 4 agents for various cases (no kidding), and trying to avoid setting a "default" agent in `.bashrc`. Right now I went to the extension source code and hacked it by hardcoding my sock path in the javascript source, rather than messing with profile.

I am not sure if it worth to continue persuading that SSH_AUTH_SOCK option is needed or just give up and keep doing some workarounds.

**Comment by maxsatula** [comment]
Also, I am not sure (hontstly, have not tested yet) whether VS Code Server runs any bash profile files upon startup. It is quite possible it may skip it overall

**Comment by bwateratmsft** [comment]
It did run `.bashrc` when I tried it, at least. I was able to get a Remote - WSL session with the Docker extension attached to an SSH Docker daemon, after killing the previously-started ssh-agent (from another terminal).

Still, I see your point about this not being clean, since it would affect the whole environment. I will keep this issue open for now. I imagine it's not a common case to mix three environments (the Windows host + WSL + remote SSH Docker daemon) but you aren't the first to try it, so we will keep it in mind. :smile:

In the meantime, I did find [this guide](https://code.visualstudio.com/docs/remote/wsl#_advanced-environment-setup-script) on how to make a script run specifically for the VSCode remote session, rather than for everything (i.e. in `.bashrc`), if that works it's probably better than `.bashrc`.

Strangely though, my experience of it running `.bashrc` when connecting contradicts this quote from that document:
> When VS Code Remote is started in WSL, no shell startup scripts are run. This was done to avoid issues with startup scripts that are tuned for shells.

**Comment by maxsatula** [comment]
`~/.vscode-server/server-env-setup` - that is probably the best solution, thank you for the link!

**Comment by lcxywfe** [comment]
@bwateratmsft Sorry, I am following the steps [here](https://code.visualstudio.com/docs/containers/choosing-dev-environment#_remote-machine), but when I wanted to "Open in Browser", I got "No valid ports were mapped from the container to the host." Do you know how to solve it?

**Comment by bwateratmsft** [comment]
Can you inspect the container and share the output?

**Comment by lcxywfe** [comment]
{
    "Id": "63e1a977986ee62b63b5c50c0066e190fc334a488e5c4703442f97b9539988db",
    "Created": "2020-02-23T00:33:35.015431438Z",
    "Path": "/bin/bash",
    "Args": [],
    "State": {
        "Status": "running",
        "Running": true,
        "Paused": false,
        "Restarting": false,
        "OOMKilled": false,
        "Dead": false,
        "Pid": 5142,
        "ExitCode": 0,
        "Error": "",
        "StartedAt": "2020-04-09T07:55:37.785094259Z",
        "FinishedAt": "2020-04-09T07:55:09.29683194Z"
    },
    "Image": "sha256:797e776a616ff3262e928f55ebd382ce798371d1cfcfbd0cf67399094d24d7e9",
    "ResolvConfPath": "/var/lib/docker/containers/63e1a977986ee62b63b5c50c0066e190fc334a488e5c4703442f97b9539988db/resolv.conf",
    "HostnamePath": "/var/lib/docker/containers/63e1a977986ee62b63b5c50c0066e190fc334a488e5c4703442f97b9539988db/hostname",
    "HostsPath": "/var/lib/docker/containers/63e1a977986ee62b63b5c50c0066e190fc334a488e5c4703442f97b9539988db/hosts",
    "LogPath": "/var/lib/docker/containers/63e1a977986ee62b63b5c50c0066e190fc334a488e5c4703442f97b9539988db/63e1a977986ee62b63b5c50c0066e190fc334a488e5c4703442f97b9539988db-json.log",
    "Name": "/dsm_inference_accelerate",
    "RestartCount": 0,
    "Driver": "overlay2",
    "Platform": "linux",
    "MountLabel": "",
    "ProcessLabel": "",
    "AppArmorProfile": "unconfined",
    "ExecIDs": null,
    "HostConfig": {
        "Binds": [
            "/home/ubuntu/Develop:/Develop"
        ],
        "ContainerIDFile": "",
        "LogConfig": {
            "Type": "json-file",
            "Config": {}
        },
        "NetworkMode": "default",
        "PortBindings": {},
        "RestartPolicy": {
            "Name": "no",
            "MaximumRetryCount": 0
        },
        "AutoRemove": false,
        "VolumeDriver": "",
        "VolumesFrom": null,
        "CapAdd": [
            "SYS_PTRACE"
        ],
        "CapDrop": null,
        "Capabilities": null,
        "Dns": [],
        "DnsOptions": [],
        "DnsSearch": [],
        "ExtraHosts": null,
        "GroupAdd": null,
        "IpcMode": "private",
        "Cgroup": "",
        "Links": null,
        "OomScoreAdj": 0,
        "PidMode": "",
        "Privileged": true,
        "PublishAllPorts": false,
        "ReadonlyRootfs": false,
        "SecurityOpt": [
            "seccomp=unconfined",
            "label=disable"
        ],
        "UTSMode": "",
        "UsernsMode": "",
        "ShmSize": 67108864,
        "Runtime": "runc",
        "ConsoleSize": [
            0,
            0
        ],
        "Isolation": "",
        "CpuShares": 0,
        "Memory": 0,
        "NanoCpus": 0,
        "CgroupParent": "",
        "BlkioWeight": 0,
        "BlkioWeightDevice": [],
        "BlkioDeviceReadBps": null,
        "BlkioDeviceWriteBps": null,
        "BlkioDeviceReadIOps": null,
        "BlkioDeviceWriteIOps": null,
        "CpuPeriod": 0,
        "CpuQuota": 0,
        "CpuRealtimePeriod": 0,
        "CpuRealtimeRuntime": 0,
        "CpusetCpus": "",
        "CpusetMems": "",
        "Devices": [],
        "DeviceCgroupRules": null,
        "DeviceRequests": [
            {
                "Driver": "",
                "Count": -1,
                "DeviceIDs": null,
                "Capabilities": [
                    [
                        "gpu"
                    ]
                ],
                "Options": {}
            }
        ],
        "KernelMemory": 0,
        "KernelMemoryTCP": 0,
        "MemoryReservation": 0,
        "MemorySwap": 0,
        "MemorySwappiness": null,
        "OomKillDisable": false,
        "PidsLimit": null,
        "Ulimits": null,
        "CpuCount": 0,
        "CpuPercent": 0,
        "IOMaximumIOps": 0,
        "IOMaximumBandwidth": 0,
        "MaskedPaths": null,
        "ReadonlyPaths": null
    },
    "GraphDriver": {
        "Data": {
            "LowerDir": "/var/lib/docker/overlay2/91309d0eb4ac96b31175ec96fb5d248217b7ff3904c60538bffded9db1c688f6-init/diff:/var/lib/docker/overlay2/8ca3bf914eb494b563e0afdac9e6add321d1c83e56fe3f6c5c11e1218a7ea45d/diff:/var/lib/docker/overlay2/971d367a678b788d80345ff739f09ba24ad007558eee3e9d45af3b6f117a02a1/diff:/var/lib/docker/overlay2/f7b3059077c3e4451826f677de3cde5f8cc30c68c6efc527f2af6c141d857317/diff:/var/lib/docker/overlay2/2ab35feb9952a94cc5deb866bc9c76d53db7c3097943c68d1292d16d106a03c3/diff:/var/lib/docker/overlay2/570f1b519d460cd88f343746ea17818009996f49e965f23f15e1ca6e492a66bf/diff:/var/lib/docker/overlay2/2b6a197b179086bffa7124df46e6cba43947bb790877afcf7488caa74c9400c0/diff:/var/lib/docker/overlay2/516d44377c6bb370bc44eac41d4e7f237e67bcc7384c9c5bc160ce911e034e0b/diff:/var/lib/docker/overlay2/59759e90e1d6553b1a85d2c4171a80a95756ac2b978f1ba244ba9e162ec3f6b3/diff:/var/lib/docker/overlay2/a9992e6d1a8e7b615dc3fbf63d993cc4af1d183bf2a347c8b4eae698f841bd75/diff:/var/lib/docker/overlay2/e8baf1df14134b4f62037357b88b2d564680da7531bc9130bf8a47bccbc002c0/diff:/var/lib/docker/overlay2/3e299122fb3f7aa4a27652476db316ca7234eab800a80f80e95de4f11f0cda6a/diff:/var/lib/docker/overlay2/86e2ac9dc5bebda12c4a628c015bc7528c8b1fff8e927062ff64bc08a2d1fc98/diff:/var/lib/docker/overlay2/6faf0bb8cf495ae3f6756e1d466ae65ef01c5ab15df8cadb931cbb0f20a0a7cf/diff:/var/lib/docker/overlay2/3d4ef16d8acfa44cdad2c026af23eac676fdf48c64d1f0ed7b23a3497696d74e/diff:/var/lib/docker/overlay2/c532017b5bd34c53da2973e7174e29bc3d87d18ecac37d53561f9f241d29659b/diff:/var/lib/docker/overlay2/b03ad69c39e91b411e343ead1c748a03452b18f5baa5a71cbcd04ef28d528806/diff:/var/lib/docker/overlay2/bcaca5d488d3b2501bf3d555a748eb9d8f47902b5cae0b80c0f080ed8c7870e0/diff:/var/lib/docker/overlay2/ed7479944c52571da9292a49fc959b6f8935371f918968e5caf93837a3030871/diff:/var/lib/docker/overlay2/4e81469b39b959449c9273f7f88b1ab93e20791cd4a7fe07323af8c23f295ff9/diff:/var/lib/docker/overlay2/fabdaed34bd999c4350961a828a4134dfb13226d64bb22e30e4febe619dc44a2/diff:/var/lib/docker/overlay2/ee33e55b50e716f8c672508eac422094e502ec50ed212263b819b0b32f9ae70a/diff:/var/lib/docker/overlay2/8ed2d46a3c03c74dd1d02b5751256e6867e0a785443e988c5eeb27aa7b60fc34/diff:/var/lib/docker/overlay2/e577ee60fb963e4a53568c5b533e59db2165eb0a525eaf2d29ab4d814165e45e/diff:/var/lib/docker/overlay2/42a35575217262072034c09ca118ce91ff3906b1130a782c51483c9abc911b4a/diff:/var/lib/docker/overlay2/3b7113f98b8d5e2dbf26a43126965266d566b42d1369853c70d7b4506976675c/diff:/var/lib/docker/overlay2/c3167569399982a2d17e5d6d2edee34957dcd0bf776a5ab310a2ef81921737f5/diff:/var/lib/docker/overlay2/aa15a45d668c2d39bca78d11108d817968728bf25901155a9d294c4616047dbe/diff:/var/lib/docker/overlay2/249ffbeb5fccd096cf9ed654346c4f656efdc25486544b7ec92a3ae7b45e1a22/diff:/var/lib/docker/overlay2/09cc0317d80e9e68258dda92550ebc2c9cb54294791ecbb860f098286d25f16d/diff:/var/lib/docker/overlay2/8e8409c0c84501432437d5b0c76f74b2bb288da76050947d6a4c35eb091dbd3e/diff:/var/lib/docker/overlay2/7c82003988b1470f172625b8284dc73f26ff0f7dba9470ce51e56c751d1a6db4/diff:/var/lib/docker/overlay2/d25628eb6a911482a074e0aaccfc2bb178cdd47a5ed94a7e2fb3def0a7fc2df7/diff:/var/lib/docker/overlay2/0f54dda71f6fb1822ae86a4c8a003da066f40b97454ebc03dd055cb1f598d9be/diff:/var/lib/docker/overlay2/c1b47f31aa0e6a98803b7c581169be9caff47e8bbee1db97898b3b6610756560/diff:/var/lib/docker/overlay2/9658d62da756d45af930fb7af4781c042eb81af7a5b55de65679631babf6abfd/diff:/var/lib/docker/overlay2/3da04b724588cd153184e8544150b06fc69cf9ac8fb3d15593383443181c971f/diff:/var/lib/docker/overlay2/3b6b2c5683a206c6fd0841e28d477393b9548d80f503e99566b21fe91f1858b7/diff:/var/lib/docker/overlay2/f96839dc96091453f11018610e09ddd013ae3cc9d335b5cdd697c692e2e1d41d/diff:/var/lib/docker/overlay2/3fbea0244a235cdd05aedb94d1d08ca01f60f77b318202b9eedeabe040ec1bd7/diff:/var/lib/docker/overlay2/af045c33ff79636f2b3d625948372449206437275d8a221dc2214a511426400a/diff:/var/lib/docker/overlay2/106c08823cdbdc32a0aa58016ef3034409cc98fb9936c88ef00d52bdcd566c78/diff:/var/lib/docker/overlay2/33ea73553675950ba151e311c52efa914d6d603fe62444c5deea8bbe3ef21116/diff:/var/lib/docker/overlay2/426aab1fbd67a83089e59d40daccf166492d574ad0e94ef72612341ca41f4650/diff:/var/lib/docker/overlay2/6b96098c37b7dbb442613126a2f244ebc4bd6871bfd4f190cae53ba9ac147e49/diff:/var/lib/docker/overlay2/94f7295c18408789a67f87f6b632e1b3894e04caee1b48ec2d5bcfa9a83cc5ee/diff:/var/lib/docker/overlay2/c7fb80ac1eac2529b1e1a548ba066115bd3123d0e0e861fb5a3c6fef3aa2ed16/diff:/var/lib/docker/overlay2/6fa66024bb4298e4e733c45e5268ab14251b4dad0cf83c9fd0a8b3f65dfcd60d/diff:/var/lib/docker/overlay2/33a69414f9dcafa5b46f817cead7e7ffbfcac4d55b4cc3a963939e1fd7ace45b/diff:/var/lib/docker/overlay2/5cec0a541419bedf67b4f4f030adb939221c4d758bd31c2b20cf39f8094af21a/diff:/var/lib/docker/overlay2/8fa7bec9b65e1e19e421363ae79df3f457090f99efa321a3fe0e23e94faeb518/diff",
            "MergedDir": "/var/lib/docker/overlay2/91309d0eb4ac96b31175ec96fb5d248217b7ff3904c60538bffded9db1c688f6/merged",
            "UpperDir": "/var/lib/docker/overlay2/91309d0eb4ac96b31175ec96fb5d248217b7ff3904c60538bffded9db1c688f6/diff",
            "WorkDir": "/var/lib/docker/overlay2/91309d0eb4ac96b31175ec96fb5d248217b7ff3904c60538bffded9db1c688f6/work"
        },
        "Name": "overlay2"
    },
    "Mounts": [
        {
            "Type": "bind",
            "Source": "/home/ubuntu/Develop",
            "Destination": "/Develop",
            "Mode": "",
            "RW": true,
            "Propagation": "rprivate"
        }
    ],
    "Config": {
        "Hostname": "63e1a977986e",
        "Domainname": "",
        "User": "",
        "AttachStdin": true,
        "AttachStdout": true,
        "AttachStderr": true,
        "Tty": true,
        "OpenStdin": true,
        "StdinOnce": true,
        "Env": [
            "PATH=/usr/local/nvidia/bin:/usr/local/cuda/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin",
            "CUDA_VERSION=10.1.243",
            "CUDA_PKG_VERSION=10-1=10.1.243-1",
            "LD_LIBRARY_PATH=/usr/local/nvidia/lib:/usr/local/nvidia/lib64",
            "NVIDIA_VISIBLE_DEVICES=all",
            "NVIDIA_DRIVER_CAPABILITIES=compute,utility",
            "NVIDIA_REQUIRE_CUDA=cuda>=10.1 brand=tesla,driver>=384,driver<385 brand=tesla,driver>=396,driver<397 brand=tesla,driver>=410,driver<411",
            "NCCL_VERSION=2.4.8",
            "LIBRARY_PATH=/usr/local/cuda/lib64/stubs",
            "CUDNN_VERSION=7.6.3.30"
        ],
        "Cmd": [
            "/bin/bash"
        ],
        "Image": "ecopia/gpu_virginia:g4-wf-inference-acc",
        "Volumes": null,
        "WorkingDir": "",
        "Entrypoint": null,
        "OnBuild": null,
        "Labels": {
            "com.nvidia.cudnn.version": "7.6.3.30",
            "maintainer": "NVIDIA CORPORATION <cudatools@nvidia.com>"
        }
    },
    "NetworkSettings": {
        "Bridge": "",
        "SandboxID": "eedbb07764e4462a975c95d2e1b56882651b2daa48c15f96e20111320f1922fe",
        "HairpinMode": false,
        "LinkLocalIPv6Address": "",
        "LinkLocalIPv6PrefixLen": 0,
        "Ports": {},
        "SandboxKey": "/var/run/docker/netns/eedbb07764e4",
        "SecondaryIPAddresses": null,
        "SecondaryIPv6Addresses": null,
        "EndpointID": "b244b90a0539bc7740909a956ce3544d0b0c442f3dd3d2cb75e575e6c9aba22e",
        "Gateway": "172.17.0.1",
        "GlobalIPv6Address": "",
        "GlobalIPv6PrefixLen": 0,
        "IPAddress": "172.17.0.2",
        "IPPrefixLen": 16,
        "IPv6Gateway": "",
        "MacAddress": "02:42:ac:11:00:02",
        "Networks": {
            "bridge": {
                "IPAMConfig": null,
                "Links": null,
                "Aliases": null,
                "NetworkID": "e44a30532262ba85ee8d077ac99ea9b913ede1c19532b1b35eceeb42668400fc",
                "EndpointID": "b244b90a0539bc7740909a956ce3544d0b0c442f3dd3d2cb75e575e6c9aba22e",
                "Gateway": "172.17.0.1",
                "IPAddress": "172.17.0.2",
                "IPPrefixLen": 16,
                "IPv6Gateway": "",
                "GlobalIPv6Address": "",
                "GlobalIPv6PrefixLen": 0,
                "MacAddress": "02:42:ac:11:00:02",
                "DriverOpts": null
            }
        }
    }
}

**Comment by bwateratmsft** [comment]
Your container does not have any ports published. How did you start it? `docker-compose up`, `docker run`, a VSCode task, etc.?

**Comment by lcxywfe** [comment]
`docker run`

**Comment by bwateratmsft** [comment]
Can you share the full command line of your `docker run` command? Most likely you need to include the `-P` option, or otherwise manually specify port publishing with `-p hostPort:containerPort`.

**Comment by lcxywfe** [comment]
` docker run  -it --gpus all -v --privileged --cap-add=SYS_PTRACE --security-opt seccomp=unconfined xxx:xxx`  Let me try the `-P`

**Comment by lcxywfe** [comment]
I added `-P`, but it still told me `No valid ports were mapped from the container to the host.`

**Comment by lcxywfe** [comment]
`-p` works, but I misunderstood the `Open in Browser1, I just want to open a folder in the container. = =

**Comment by bwateratmsft** [comment]
Ah, gotcha. The Docker extension doesn't have a file browser for containers. The [Remote - Containers](https://marketplace.visualstudio.com/items?itemName=ms-vscode-remote.remote-containers) can sorta do it, if you attach VSCode to the container, and do "Open Folder", you can see things.

Honestly the easiest way is to attach a shell to the container and use the command line to explore.

**Comment by lcxywfe** [comment]
Thanks,  because my development environment is in a container : )

**Comment by bwateratmsft** [comment]
@steverice we've been looking at a way to do bring-your-own-Dockerode-settings to enable users to have complete control over what Dockerode does. I see that line you linked where the options are passed into `ssh2`, but it looks like based on [this line](https://github.com/apocas/docker-modem/blob/master/lib/modem.js#L210), `docker-modem` does not pass all its input options into that function. So, in the current version of `docker-modem`, we still couldn't also fully control the options to `ssh2`. Additionally, since both `docker-modem` and `ssh2` use the `agent` property but for different meanings, it could be difficult to control the SSH agent thing.

Is the need strong enough to make an issue/PR in `docker-modem`?

**Comment by steverice** [comment]
@bwateratmsft I'm not exactly sure 😁 
What I'm trying to do is [run a devcontainer using Docker on a remote host](https://code.visualstudio.com/docs/remote/containers-advanced#_developing-inside-a-container-on-a-remote-docker-host).

This seems to work okay after playing with some settings (and using `docker context`) except that the remote machine is lacking permissions to pull the docker images it needs from our internal ECR repo. I believe this works ordinarily from the CLI because of the following in my `~/.ssh/config` (on macOS):
```
  ForwardAgent yes
  UseKeychain yes
  AddKeysToAgent yes
```

I came across this issue because it seems to deal with the `docker-modem`/`dockerode` settings issue… and I gathered from elsewhere that remote-container is using the code here in `vscode-docker` for communication. I'll try and hack an [`agentForward: true`](https://github.com/mscdex/ssh2#client-methods) into the docker-modem code to see if it fixes my problem and report back.

(this is also why the line you linked wouldn't be an issue for me, as it only overrides the `agent` option)

**Comment by bwateratmsft** [comment]
It overrides the `agent` option but also it _doesn't_ pass all of its options into `ssh()`; only the host/port/username/password/agent. So if you gave `agentForward: true` to Dockerode it still wouldn't be passed all the way to `ssh2` :(

**Comment by karolz-ms** [comment]
> @steverice we've been looking at a way to do bring-your-own-Dockerode-settings to enable users to have complete control over what Dockerode does. I see that line you linked where the options are passed into `ssh2`, but it looks like based on [this line](https://github.com/apocas/docker-modem/blob/master/lib/modem.js#L210), `docker-modem` does not pass all its input options into that function. 

@bwateratmsft my understanding is `Object.assign` will only add to the target object, so whatever was passed in will be retained. Am I missing something?

**Comment by steverice** [comment]
@karolz-ms I think you're missing the same thing I did, which is that the options we care about are in the call to `ssh()` on that line and that line alone.
The `Object.assign` is being used for the `http` options which are used later.

I've opened https://github.com/apocas/docker-modem/pull/118, let's see how that does.

**Comment by bwateratmsft** [comment]
That's correct, but the `Object.assign` is docker-modem setting its own options and overriding `agent` and `protocol`. It doesn't pass `options` into `ssh()`, so the settings from on high do not make it all the way to `ssh2`.

**Comment by karolz-ms** [comment]
Roger, thanks guys

**Comment by bwateratmsft** [comment]
I've added a new setting, `docker.dockerodeOptions`, which will override everything else, and is passed unconditionally to the Dockerode constructor.

**Comment by bwateratmsft** [comment]
The fix is now available in version 1.2.0 of the [Docker extension](https://marketplace.visualstudio.com/items?itemName=ms-azuretools.vscode-docker).
